src/kernels/composable_kernel/include/kernel_algorithm/gridwise_generic_2d_reduction_blockwise.hpp

```diff
@@ -0,0 +1,462 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef CK_GRIDWISE_GENERIC_2D_REDUCTION_BLOCKWISE_HPP
+#define CK_GRIDWISE_GENERIC_2D_REDUCTION_BLOCKWISE_HPP
+
+#include "float_type.hpp"
+#include "reduction_operator.hpp"
+#include "reduction_functions.hpp"
+#include "reduction_common.hpp"
+
+#include "blockwise_generic_tensor_slice_copy.hpp"
+#include "ConstantMatrixDescriptor.hpp"
+
+namespace ck {
+
+template <int BlockSize,
+          typename srcDataType,
+          typename dstDataType,
+          typename src2dDesc,
+          typename dst1dDesc,
+          typename compType,
+          ckReduceTensorOp_t op,
+          ckNanPropagation_t nanPropaOpt,
+          ckReduceTensorIndices_t reduceIndicesOpt,
+          int callId,
+          int GredAccessesPerThreadInBlock>
+struct Gridwise_generic_reduction_xy_to_x_blockwise
+{
+    static constexpr bool indexable = reduce_binary_operator<compType, op>::indexable;
+    static constexpr bool need_indices =
+        indexable && (reduceIndicesOpt != CK_REDUCE_TENSOR_NO_INDICES);
+    static constexpr bool firstCall = (callId == 0) ? true : false;
+
+    static constexpr int BlockBufferSize = BlockSize * GredAccessesPerThreadInBlock;
+
+    using opReduce = typename reduce_binary_operator<compType, op>::opType;
+
+    __device__ void Run(srcDataType alpha,
+                        const srcDataType* const __restrict__ p_src_global,
+                        dstDataType beta,
+                        dstDataType* const __restrict__ p_dst_global,
+                        int* const __restrict__ ws_indices_global,
```

Comment:
This is the index type decided by framework, so you don't need to use index_t

But are they always ```int``` type? If not, should you make it templated?

Reply:
I could be other type,  but so far cudnnReduceTensor() also only supports int32 as reduced indices.

Reply:
Support types other than int32 as reduce indices is expensive while has little benefit.